File: frontend/src/api/agent.ts
export interface ChatMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface ToolCallInfo {
  tool_name: string
  arguments: Record<string, any>
  result: Record<string, any>
  error: string
  round: number
}

/** 引用来源（仅联网搜索结果，后端已校验原文真实性） */
export interface Citation {
  ref_id: number
  quote: string
  source_type: 'web_search'
  title: string
  url: string
}

export interface AgentQueryRequest {
  query: string
  system_prompt?: string
  history?: ChatMessage[]
}

export interface AgentQueryResponse {
  answer: string
  warning_level: 'I' | 'II' | 'III' | 'IV' | ''
  reasoning: string
  actions: string[]
  tool_calls: ToolCallInfo[]
  citations: Citation[]
  rounds: number
  intent: 'chitchat' | 'agent_task'
}

// ====== SSE 流式接口 ======

/** 推理步骤 */
export interface ReasoningStep {
  step: 'router' | 'planner' | 'executor' | 'reflector' | 'synthesizer' | 'direct_chat'
  phase: 'start' | 'thinking' | 'decision' | 'done'
  message: string
  details?: Record<string, any>
}

/** 综合研判结构化结论（在 answer 流式输出前推送） */
export interface SynthMeta {
  warning_level: 'I' | 'II' | 'III' | 'IV' | ''
  reasoning: string
  actions: string[]
  citations?: Citation[]
}

/** SSE 事件类型 */
export interface AgentStreamEvent {
  type:
    | 'reasoning_step'
    | 'intent'
    | 'tool_call'
    | 'tool_result'
    | 'synth_meta'
    | 'answer_delta'
    | 'done'
    | 'error'
  // reasoning_step
  step?: ReasoningStep['step']
  phase?: ReasoningStep['phase']
  message?: string
  details?: Record<string, any>
  // intent
  intent?: 'chitchat' | 'agent_task'
  // tool_call / tool_result
  tool?: string
  arguments?: Record<string, any>
  result?: Record<string, any>
  error?: string
  round?: number
  // synth_meta
  data?: AgentQueryResponse | SynthMeta
  // answer_delta
  content?: string
  // error
}

/**
 * Agent 流式查询（SSE）
 *
 * 使用 fetch + ReadableStream 解析 text/event-stream。
 * onEvent 每收到一个事件就被回调一次。
 *
 * P5 改进：
 * - 总超时 5 分钟（Agent 最长耗时预估 < 3 分钟，留余量）
 * - 静默超时 60 秒（无任何事件包括心跳时触发，提示连接异常）
 * - 后端每 15s 发 :keep-alive 心跳，前端解析时跳过 comment 行
 *
 * @returns AbortController 可用于中断请求
 */
export function queryAgentStream(
  req: AgentQueryRequest,
  onEvent: (event: AgentStreamEvent) => void,
  onError?: (err: Error) => void,
): AbortController {
  const controller = new AbortController()
  const baseURL = import.meta.env.VITE_API_BASE_URL || '/api'

  // 总超时：5 分钟（覆盖 planner+executor+synthesizer 全链路 + 余量）
  const TOTAL_TIMEOUT_MS = 5 * 60 * 1000
  // 静默超时：60 秒无任何数据（含心跳）则判定连接异常
  const SILENCE_TIMEOUT_MS = 60 * 1000

  let silenceTimer: ReturnType<typeof setTimeout> | null = null
  let totalTimer: ReturnType<typeof setTimeout> | null = null
  // 是否已收到终止事件（done/error）。流正常结束（EOF）但未收到终止事件时视为连接中断
  let receivedTerminal = false

  const clearTimers = () => {
    if (silenceTimer) clearTimeout(silenceTimer)
    if (totalTimer) clearTimeout(totalTimer)
    silenceTimer = null
    totalTimer = null
  }

  const resetSilenceTimer = () => {
    if (silenceTimer) clearTimeout(silenceTimer)
    silenceTimer = setTimeout(() => {
      console.warn('[SSE] 静默超时，中断连接')
      controller.abort()
      if (onError) onError(new Error('SSE 连接静默超时（60 秒无数据）'))
    }, SILENCE_TIMEOUT_MS)
  }

  totalTimer = setTimeout(() => {
    console.warn('[SSE] 总超时，中断连接')
    controller.abort()
    if (onError) onError(new Error('Agent 运行超时（5 分钟）'))
  }, TOTAL_TIMEOUT_MS)

  // 使用 fetch 而非 axios，因为 axios 不原生支持 stream
  fetch(`${baseURL}/agent/query/stream`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
    signal: controller.signal,
  })
    .then(async (response) => {
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`)
      }
      if (!response.body) {
        throw new Error('响应无内容（body 为空）')
      }
      const reader = response.body.getReader()
      const decoder = new TextDecoder('utf-8')
      let buffer = ''

      // 收到首包，启动静默计时
      resetSilenceTimer()

      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        // 任意数据到达，重置静默计时
        resetSilenceTimer()
        buffer += decoder.decode(value, { stream: true })

        // SSE 事件以 \n\n 分隔
        const events = buffer.split('\n\n')
        buffer = events.pop() || '' // 保留最后未完整的事件

        for (const rawEvent of events) {
          const line = rawEvent.trim()
          // P5：跳过后端心跳 comment（: 开头的行）
          if (!line.startsWith('data:')) continue
          const jsonStr = line.slice(5).trim()
          if (!jsonStr) continue
          try {
            const event = JSON.parse(jsonStr) as AgentStreamEvent
            onEvent(event)
            // done/error 事件后清理计时器
            if (event.type === 'done' || event.type === 'error') {
              receivedTerminal = true
              clearTimers()
            }
          } catch (e) {
            console.warn('[SSE] 解析事件失败:', jsonStr, e)
          }
        }
      }
      clearTimers()
      // 流正常结束（EOF）但始终未收到 done/error：服务端异常关闭，
      // 主动触发 onError 兜底，避免上层 loading 永久卡住
      if (!receivedTerminal) {
        console.warn('[SSE] 流结束但未收到终止事件（done/error），判定连接中断')
        if (onError) onError(new Error('连接中断，未收到完整响应'))
      }
    })
    .catch((err) => {
      clearTimers()
      if (err.name === 'AbortError') return
      console.error('[SSE] 请求失败:', err)
      if (onError) onError(err)
    })

  return controller
}

File: backend/app/api/agent.py
"""Agent 对话接口。"""
import json
import logging
import threading
import time
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from agent.graph.workflow import LLMError, run_graph_agent, run_graph_agent_stream_v2

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/agent", tags=["agent"])

# SSE 心跳间隔（秒）。长连接无数据时定期发送 comment 保活，
# 避免代理/负载均衡提前断开（参考 nginx proxy_read_timeout 默认 60s）
SSE_KEEPALIVE_INTERVAL = 15.0


# ====== 请求 / 响应模型 ======

class ChatMessage(BaseModel):
    """对话消息。"""

    role: str = Field(..., description="user/assistant")
    content: str = Field(..., description="消息内容")


class AgentQueryRequest(BaseModel):
    """Agent 查询请求。"""

    query: str = Field(..., min_length=1, max_length=2000, description="用户问题")
    system_prompt: Optional[str] = Field(None, description="自定义系统提示词（暂未使用，预留）")
    history: List[ChatMessage] = Field(
        default_factory=list, description="历史对话（不含当前问题）"
    )


class ToolCallInfo(BaseModel):
    """工具调用信息。"""

    tool_name: str
    arguments: Dict[str, Any]
    result: Dict[str, Any]
    error: str = ""
    round: int = 1


class AgentQueryResponse(BaseModel):
    """Agent 查询响应（LangGraph 版本）。"""

    answer: str = Field(..., description="最终自然语言回答")
    warning_level: str = Field("", description="预警等级 I/II/III/IV；闲聊为空")
    reasoning: str = Field("", description="研判依据；闲聊为空")
    actions: List[str] = Field(default_factory=list, description="应急措施列表")
    tool_calls: List[ToolCallInfo] = Field(
        default_factory=list, description="工具调用链路"
    )
    rounds: int = Field(0, description="实际执行的轮次")
    intent: str = Field("agent_task", description="意图：chitchat / agent_task")


# ====== 接口 ======

@router.post("/query", response_model=AgentQueryResponse)
def agent_query(req: AgentQueryRequest) -> AgentQueryResponse:
    """接收用户问题，运行 LangGraph Agent，返回预警研判结果。"""
    logger.info("Agent query (graph): %s", req.query[:200])

    try:
        history = [m.model_dump() for m in req.history]
        result = run_graph_agent(user_query=req.query, history=history)
    except LLMError as e:
        # P2：LLM 异常按分类返回对应 HTTP 状态码
        logger.warning("[agent_query] LLM error: %s (kind=%s)", e, e.kind)
        raise HTTPException(status_code=e.status_code, detail=str(e)) from e
    except Exception as e:
        logger.exception("Agent run failed")
        raise HTTPException(status_code=500, detail=f"Agent 运行失败: {e}") from e

    return AgentQueryResponse(
        answer=result["final_answer"],
        warning_level=result["warning_level"],
        reasoning=result["reasoning"],
        actions=result["actions"],
        tool_calls=[
            ToolCallInfo(
                tool_name=tc["tool_name"],
                arguments=tc["arguments"],
                result=tc["result"],
                error=tc.get("error", ""),
                round=tc.get("round", 1),
            )
            for tc in result["tool_calls"]
        ],
        rounds=result["rounds"],
        intent=result.get("intent", "agent_task"),
    )


# ====== SSE 流式接口 ======

def _sse_event(data: dict) -> str:
    """格式化为 SSE 事件字符串。"""
    return f"data: {json.dumps(data, ensure_ascii=False)}\n\n"


def _sse_keepalive_comment() -> str:
    """SSE comment 心跳（客户端忽略，仅用于保活）。"""
    return ": keep-alive\n\n"


def _stream_generator(query: str, history: list):
    """SSE 事件生成器（P5 加心跳保活）。

    使用独立线程跑 Agent，主线程定期发心跳避免连接超时。
    """
    # 用 queue 把 Agent 事件传到主线程
    import queue
    event_queue: "queue.Queue[Any]" = queue.Queue()
    # 哨兵对象标记 Agent 结束
    _SENTINEL = object()

    def _agent_worker():
        try:
            for event in run_graph_agent_stream_v2(user_query=query, history=history):
                event_queue.put(event)
        except LLMError as e:
            # P2：LLM 异常封装为 error 事件推给前端
            logger.warning("[sse] LLM error: %s (kind=%s)", e, e.kind)
            event_queue.put({"type": "error", "message": str(e),
                             "kind": e.kind, "status_code": e.status_code})
        except Exception as e:
            logger.exception("[sse] 流式生成失败")
            event_queue.put({"type": "error", "message": str(e)})
        finally:
            event_queue.put(_SENTINEL)

    worker_thread = threading.Thread(target=_agent_worker, daemon=True)
    worker_thread.start()

    last_emit = time.time()
    try:
        while True:
            try:
                # 最多等 keepalive 间隔，超时就发心跳
                item = event_queue.get(timeout=max(0.1, SSE_KEEPALIVE_INTERVAL - (time.time() - last_emit)))
            except queue.Empty:
                # 发心跳
                yield _sse_keepalive_comment()
                last_emit = time.time()
                continue

            if item is _SENTINEL:
                break
            yield _sse_event(item)
            last_emit = time.time()
    finally:
        if worker_thread.is_alive():
            # 主线程退出时 worker 仍可能在跑，daemon=True 会随进程退出
            logger.warning("[sse] stream ended before agent finished, worker still running")


@router.post("/query/stream")
def agent_query_stream(req: AgentQueryRequest):
    """Agent 对话流式接口（SSE）。

    返回 text/event-stream，事件类型：
      - reasoning_step：推理步骤（router/planner/executor/reflector/synthesizer/direct_chat）
      - intent：意图识别结果
      - tool_call / tool_result：工具调用链
      - synth_meta：综合研判结构化结论（warning_level/reasoning/actions）
      - answer_delta：token 级流式答案
      - done：完整响应
      - error：错误

    每条事件格式：data: {"type": "...", ...}\n\n
    """
    logger.info("Agent query (stream v2): %s", req.query[:200])
    history = [m.model_dump() for m in req.history]
    return StreamingResponse(
        _stream_generator(req.query, history),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",  # 禁用 nginx 缓冲
        },
    )
